File: tests/popart/opset_check_test.py
import torch
import torch.nn as nn
import torch.nn.functional as F
import popart
import onnx
import test_util as tu
import pytest

resize_op_error = (
    "For an opset 11 Model, the ONNX spec stipulates that a "
    "Resize op must be version 11. The highest version we "
    "have implemented less than or equal to 11 is 10, so bailing.")


def test_opset():
    class X(nn.Module):
        def __init__(self):
            super(X, self).__init__()

        def forward(self, x):
            return F.upsample(x, scale_factor=[2, 2])

    model = X()
    data = torch.randn(1, 1, 3, 3)
    torch.onnx.export(model, data, 'foo.onnx', verbose=True, opset_version=11)

    model = onnx.load_model('foo.onnx')
    # This test only works if there is a resize in the model
    assert 'Resize' in [node.op_type for node in model.graph.node]

    dataFlow = popart.DataFlow(
        1, {model.graph.output[0].name: popart.AnchorReturnType("All")})
    with pytest.raises(popart.popart_exception) as e_info:
        sess = popart.InferenceSession(fnModel='foo.onnx',
                                       dataFlow=dataFlow,
                                       deviceInfo=tu.create_test_device())

    assert resize_op_error in str(e_info.value)


def test_strictOpVersions():
    def run_test(strictOpVersions):
        class X(nn.Module):
            def __init__(self):
                super(X, self).__init__()

            def forward(self, x):
                return F.upsample(x, scale_factor=[2, 2])

        model = X()
        data = torch.randn(1, 1, 3, 3)
        torch.onnx.export(model,
                          data,
                          'foo.onnx',
                          verbose=True,
                          opset_version=11)

        model = onnx.load_model('foo.onnx')
        # This test only works if there is a resize in the model
        assert 'Resize' in [node.op_type for node in model.graph.node]

        dataFlow = popart.DataFlow(
            1, {model.graph.output[0].name: popart.AnchorReturnType("All")})
        opts = popart.SessionOptions()
        opts.strictOpVersions = strictOpVersions

        try:
            sess = popart.InferenceSession(fnModel='foo.onnx',
                                           dataFlow=dataFlow,
                                           deviceInfo=tu.create_test_device(),
                                           userOptions=opts)
        except popart.popart_exception as e:
            return str(e)

        return ""

    # Check that the version error is only returned when strictOpVersions is True
    assert resize_op_error in run_test(True)
    assert resize_op_error not in run_test(False)

File: tests/popart/opset_check_test.cpp
// Copyright (c) 2018 Graphcore Ltd. All rights reserved.
#define BOOST_TEST_MODULE OpsetCheckTest

#include <boost/test/unit_test.hpp>

#include <popart/builder.hpp>
#include <popart/dataflow.hpp>
#include <popart/devicemanager.hpp>
#include <popart/opidentifier.hpp>
#include <popart/opmanager.hpp>
#include <popart/session.hpp>

using namespace popart;

class DepthToSpaceOp : public Op {
public:
  DepthToSpaceOp(const OperatorIdentifier &opid_, const Op::Settings &settings_)
      : Op(opid_, settings_) {}

  std::unique_ptr<Op> clone() const override { throw error("unimplemented"); }
  float getSubgraphValue() const override { throw error("unimplemented"); }
  void setup() override { throw error("setup is unimplemented"); }
};

static OpDefinition depthToSpaceOpDef({OpDefinition::Inputs({}),
                                       OpDefinition::Outputs({}),
                                       OpDefinition::Attributes({})});

static OpCreator<DepthToSpaceOp> depthToSpaceOpCreator(OpDefinitions({
    {Onnx::Operators::DepthToSpace_1, depthToSpaceOpDef},

}));

// This test works because DepthToSpace is not implemented in any version at the
// moment. If DepthToSpace is ever implemented, this op will break.
//
// This test checks that the strict onnx opset version checking works, and also
// checks that the session option 'strictOpVersions' works.
BOOST_AUTO_TEST_CASE(OpsetCheck) {
  auto test = [](bool strictOpsetVersions) {
    auto builder     = Builder::create();
    auto aiGraphcore = builder->aiGraphcoreOpset1();

    TensorInfo inputInfo{"FLOAT", std::vector<int64_t>{2}};

    auto input = builder->addInputTensor(inputInfo);
    std::vector<TensorId> outputs;
    if (strictOpsetVersions) {
      outputs = builder->customOp(
          Onnx::Operators::DepthToSpace_1, 11, {input}, 1, {});
    } else {
      outputs =
          builder->customOp(Onnx::Operators::DepthToSpace_1, 9, {input}, 1, {});
    }

    auto l1 = aiGraphcore.l1loss({outputs.at(0)}, 0.1f, ReductionType::Sum);

    auto proto = builder->getModelProto();

    auto dataFlow = DataFlow(1, {{outputs.at(0), AnchorReturnType("All")}});

    auto cpuDevice = DeviceManager::createDeviceManager().createCpuDevice();

    SessionOptions opts   = SessionOptions();
    opts.strictOpVersions = strictOpsetVersions;

    // Create the session
    BOOST_CHECK_EXCEPTION(
        InferenceSession::createFromOnnxModel(
            proto, dataFlow, cpuDevice, {}, opts),
        error,
        [&](error const &err) {
          std::string what    = err.what();
          auto expected_error = [&]() -> std::string {
            if (strictOpsetVersions) {
              return "For an opset 11 Model, the ONNX spec stipulates that a "
                     "DepthToSpace op must be version 11. The highest version "
                     "we have "
                     "implemented less than or equal to 11 is 1, so bailing.";
            } else {
              return "setup is unimplemented";
            }
          };

          return what == expected_error();
        });
  };

  test(true);
  test(false);
}
